docs/docs/contributing/debugging.md

```diff
@@ -0,0 +1,332 @@
+# Debugging Techniques
+- [Setting up the playground](#setting-up-the-playground)
+- [Show for human readable output](#show-for-human-readable-output)
+- [How to disable color](#how-to-disable-color)
+- [Reporting as a non-intrusive println](#reporting-as-a-non-intrusive-println)
+- [Printing out trees after phases](#printing-out-trees-after-phases)
+- [Configuring the printer output](#configuring-the-printer-output)
+- [Figuring out an object creation site](#figuring-out-an-object-creation-site)
+  * [Via ID](#via-id)
+  * [Via tracer](#via-tracer)
+- [Built-in Logging Architecture](#built-in-logging-architecture)
+  * [Printers](#printers)
+  * [Tracing](#tracing)
+  * [Reporter](#reporter)
+
+<small><i><a href='http://ecotrust-canada.github.io/markdown-toc/'>Table of contents generated with markdown-toc</a></i></small>
+
+## Setting up the playground
+Consider the `../issues/Playground.scala` (relative to the Dotty directory) file is:
+
+```scala
+object Playground {
+  def main(args: Array[String]) = {
+    println("Hello World")
+  }
+}
+```
+
+Then, you can debug Dotty by compiling this file via `dotc ../issues/Playground.scala` (from the SBT console) and collecting various debug output in process. This section documents techniques you can use to collect the debug info.
+
+[This](https://github.com/lampepfl/dotty/blob/10526a7d0aa8910729b6036ee51942e05b71abf6/compiler/src/dotty/tools/dotc/typer/Typer.scala#L2231) is the entry point to the Typer. The job of the Typer is to take an untyped tree, compute its type and turn it into a typed tree by attaching the type information to that tree (in an immutable way of course). We will use this entry point to practice debugging techniques. E.g.:
```

Comment:
> (in an immutable way of course)

Technically, we do use mutation to set types on trees: https://github.com/lampepfl/dotty/blob/dc0f865988b2abe84345bf63213087c8af5809f2/compiler/src/dotty/tools/dotc/ast/Trees.scala#L73-L79, so I would remove this comment.